### PR TITLE
schedule: support regions split

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: Song Gao <disxiaofei@163.com>

### What problem does this PR solve?

ref https://github.com/tikv/pd/issues/2965  support region splitter in `schedule` pkg

### Check List

Tests

- Unit test

### CODE REVIEW

File: server/schedule/region_splitter_test.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+type testRegionSplitterSuite struct{}
+
+func (s *testRegionSplitterSuite) TestRegionSplitter(c *C) {
+	opt := config.NewTestOptions()
+	tc := mockcluster.NewCluster(opt)
+	handler := newMockSplitRegionsHandler()
+	tc.AddLeaderRegionWithRange(1, "aaa", "zzz", 2, 3)
+	splitter := NewRegionSplitter(tc, handler)
+	newRegions := map[uint64]struct{}{}
+	failureKeys := splitter.splitRegionsByKeys([][]byte{[]byte("bbb"), []byte("ccc")}, newRegions)
```
Comment (rleungx) [inline server/schedule/region_splitter_test.go:76]: Would you like to add some cases that there are some keys which are failed to split?

Comment (Yisaer) [reply server/schedule/region_splitter_test.go]: updated.

File: server/schedule/region_splitter_test.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	// regionID -> startKey, endKey
+	regions map[uint64][2][]byte
+}
+
+func newMockSplitRegionsHandler() *mockSplitRegionsHandler {
+	return &mockSplitRegionsHandler{
+		groupKeysByRegion: map[uint64][][]byte{},
+		regions:           map[uint64][2][]byte{},
+	}
+
```
Comment (rleungx) [inline server/schedule/region_splitter_test.go:38]: remove this line

File: server/schedule/region_splitter.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		}
+		// filter start key
+		if bytes.Equal(region.GetStartKey(), key) {
+			continue
+		}
+		group, ok := groupKeys[region.GetID()]
+		if !ok {
+			groupKeys[region.GetID()] = [][]byte{}
+		}
+		//TODO: log reduction
```
Comment (rleungx) [inline server/schedule/region_splitter.go:110]: Why not directly change it?

File: server/schedule/region_splitter.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+// NewRegionSplitter return a region splitter
+func NewRegionSplitter(cluster opt.Cluster, handler SplitRegionsHandler) *RegionSplitter {
+	return &RegionSplitter{
+		cluster: cluster,
+		handler: handler,
+	}
+}
+
+// SplitRegions support splitRegions by given split keys.
+func (r *RegionSplitter) SplitRegions(splitKeys [][]byte, retryLimit int) (int, []uint64) {
```
Comment (rleungx) [inline server/schedule/region_splitter.go:48]: It's better to add tests for it?

File: server/schedule/region_splitter.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+			break
+		}
+		//TODO: sleep for a while
+		time.Sleep(500 * time.Millisecond)
+	}
+	returned := make([]uint64, 0, len(newRegions))
+	for regionID := range newRegions {
+		returned = append(returned, regionID)
+	}
+	return 100 - len(unprocessedKeys)*100/len(splitKeys), returned
```
Comment (rleungx) [inline server/schedule/region_splitter.go:63]: Will the splitKeys be 0?

Comment (Yisaer) [reply server/schedule/region_splitter.go:67]: updated.

File: server/schedule/region_splitter.go (hunk 4 above)
Comment (Yisaer) [reply server/schedule/region_splitter.go:49]: updated.

File: server/schedule/region_splitter.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+// SplitRegions support splitRegions by given split keys.
+func (r *RegionSplitter) SplitRegions(splitKeys [][]byte, retryLimit int) (int, []uint64) {
+	if len(splitKeys) < 1 {
+		return 0, nil
+	}
+	unprocessedKeys := splitKeys
+	newRegions := make(map[uint64]struct{}, len(splitKeys))
+	for i := 0; i < retryLimit; i++ {
```
Comment (rleungx) [inline server/schedule/region_splitter.go:55]: should be `i <= retryLimit`?

Comment (Yisaer) [reply server/schedule/region_splitter.go]: updated.

File: server/schedule/region_splitter.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+// GroupKeysByRegion separates keys into groups by their belonging Regions.
+// If any key failed to be found its Region, it will be placed into unProcessed key.
+// If the key is exactly the start key of its region, the key would be discarded directly.
+func (r *RegionSplitter) groupKeysByRegion(keys [][]byte) (map[uint64][][]byte, [][]byte) {
+	unProcessedKeys := make([][]byte, 0, len(keys))
+	groupKeys := make(map[uint64][][]byte, len(keys))
+	for _, key := range keys {
+		region := r.cluster.GetRegionByKey(key)
+		if region == nil {
```
Comment (rleungx) [inline server/schedule/region_splitter.go:104]: Maybe retry cannot solve this problem?

File: server/schedule/region_splitter.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+			continue
+		}
+		group, ok := groupKeys[region.GetID()]
+		if !ok {
+			groupKeys[region.GetID()] = [][]byte{}
+		}
+		log.Info("found region",
+			zap.Uint64("regionID", region.GetID()),
+			logutil.ZapRedactByteString("key", key))
+		groupKeys[region.GetID()] = append(group, key)
```
Comment (disksing) [inline server/schedule/region_splitter.go:120]: I think we can `groupKeys[region.GetID()] = append(groupKeys[region.GetID()] , key)` directly

File: server/schedule/region_splitter.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+func (r *RegionSplitter) splitRegionsByKeys(splitKeys [][]byte, newRegions map[uint64]struct{}) [][]byte {
+	//TODO: support batch limit
+	groupKeys, unProcessedKeys := r.groupKeysByRegion(splitKeys)
+	for regionID, keys := range groupKeys {
+		region := r.cluster.GetRegion(regionID)
+		// TODO: assert region is not nil
+		// TODO: assert leader exists
+		// TODO: assert region replicated
+		// TODO: assert region not hot
```
Comment (disksing) [inline server/schedule/region_splitter.go:80]: too many todos= =

File: server/schedule/region_splitter_test.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+	tc.AddLeaderRegionWithRange(2, "ccc", "eee", 2, 3)
+	tc.AddLeaderRegionWithRange(3, "fff", "ggg", 2, 3)
+	splitter := NewRegionSplitter(tc, handler)
+	groupKeys, unprocessKeys := splitter.groupKeysByRegion([][]byte{
+		[]byte("bbb"),
+		[]byte("ddd"),
+		[]byte("fff"),
+		[]byte("zzz"),
+	})
+	c.Assert(len(groupKeys), Equals, 2)
```
Comment (disksing) [inline server/schedule/region_splitter_test.go:108]: better to check in detail, not only length

File: server/schedule/region_splitter_test.go [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+func (s *testRegionSplitterSuite) TestRegionSplitter(c *C) {
+	opt := config.NewTestOptions()
+	tc := mockcluster.NewCluster(opt)
+	handler := newMockSplitRegionsHandler()
+	tc.AddLeaderRegionWithRange(1, "eee", "hhh", 2, 3)
+	splitter := NewRegionSplitter(tc, handler)
+	newRegions := map[uint64]struct{}{}
+	// assert success
+	failureKeys := splitter.splitRegionsByKeys([][]byte{[]byte("fff"), []byte("ggg")}, newRegions)
+	c.Assert(len(failureKeys), Equals, 0)
```
Comment (disksing) [inline server/schedule/region_splitter_test.go:77]: I think we can call some `GetRegionByKey` here to verify new region's start key and end key.

File: server/schedule/region_splitter.go (hunk 9 above)
Comment (Yisaer) [reply server/schedule/region_splitter.go:80]: I use TODO here in order not to make this PR too big. The detailed implement could be referred here https://github.com/tikv/pd/pull/3072.

Comment (disksing) [reply server/schedule/region_splitter.go:80]: ok

File: server/schedule/region_splitter_test.go (hunk 11 above)
Comment (Yisaer) [reply server/schedule/region_splitter_test.go:77]: We can't verify the new regions key as the `SplitRegionsHandler` is mocked during the test.

File: server/schedule/region_splitter_test.go (hunk 10 above)
Comment (Yisaer) [reply server/schedule/region_splitter_test.go:108]: updated.

File: server/schedule/region_splitter.go (hunk 8 above)
Comment (Yisaer) [reply server/schedule/region_splitter.go]: updated.

File: server/schedule/region_splitter.go [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+func (r *RegionSplitter) SplitRegions(splitKeys [][]byte, retryLimit int) (int, []uint64) {
+	if len(splitKeys) < 1 {
+		return 0, nil
+	}
+	unprocessedKeys := splitKeys
+	newRegions := make(map[uint64]struct{}, len(splitKeys))
+	for i := 0; i <= retryLimit; i++ {
+		if i > 0 {
+			//TODO: sleep for a while
+			time.Sleep(500 * time.Millisecond)
```
Comment (disksing) [inline server/schedule/region_splitter.go:58]: Can the TODO line be removed? And I think we can move it to the end of the loop so we don't have to check `i`.

**Review by disksing — APPROVED** [review]
the rest LGTM.

**Pull request merged**